### ISSUE TITLE
Reorganize tests to save compile time

### ISSUE THREAD

**Issue by mzhangmzz** [near/nearcore#5371]
Currently, many tests in the code are located in a separate crate, which hurts compile time. For example, `chain/client/tests`. We should rethink unit tests vs integration tests and move unit tests to within the crate, for example, move `chain/client/tests` to `chain/client/src/tests`, as per [https://matklad.github.io/2021/02/27/delete-cargo-integration-tests.html](https://matklad.github.io/2021/02/27/delete-cargo-integration-tests.html)

*   <span class="error">[x]</span> [https://github.com/near/nearcore/pull/5398](https://github.com/near/nearcore/pull/5398)
*   <span class="error">[x]</span> [https://github.com/near/nearcore/pull/5420](https://github.com/near/nearcore/pull/5420)
*   <span class="error">[x]</span> [https://github.com/near/nearcore/pull/5493](https://github.com/near/nearcore/pull/5493)
*   <span class="error">[x]</span> [https://github.com/near/nearcore/pull/5494](https://github.com/near/nearcore/pull/5494)
*   [ ] TODO /core
*   [ ] TODO /runtime
*   [ ] TODO rest

**Comment by mzhangmzz** [comment]
cc @matklad

**Comment by mzhangmzz** [comment]
related issue https://github.com/near/nearcore/issues/5149

**Comment by pmnoxx** [comment]
> Currently, many tests in the code are located in a separate crate, which hurts compile time. For example, `chain/client/tests`. We should rethink unit tests vs integration tests and move unit tests to within the crate, for example, move `chain/client/tests` to `chain/client/src/tests`, as per https://matklad.github.io/2021/02/27/delete-cargo-integration-tests.html

@mzhangmzz I agree. There are other benefits to doing that as well. I often looks at tests within the module for examples on how the code is used. This makes it easier to understand the code.

**Comment by pmnoxx** [comment]
@mzhangmzz  Based on  https://matklad.github.io/2021/02/27/delete-cargo-integration-tests.html
I see that for each crate, `(1+N)` binaries will be created, were
- 1 : binary will all tests inside `/src`
- N:  1 for each module in `/tests/`

We have 3 proposals:
A) keep as it is - (1+N) binaries needs to be build for tests. 
Those tests will only use public api.
B) move tests from `/tests` to `/tests/integration` - We need to build 2 binaries for tests.
Those tests will only use public api.
C) move tests from `/tests` to `/src/tests` - We need to build 1 binaries for tests.
Those tests will be able to use private API api.

The document you mentioned proposes method `B`, you are proposing method `C`.

Obviously, method `A` is suboptimal, due to linking time being slow in rust.
Method `B` is obviously superior to `A` with not side effects.

Method `C` will expose privates of the crate to those tests. That's not necessary a bad thing, but something to keep in mind.
Also build time of `C` may theoretically make regular build slower, as there are more files to parse for regular build. I don't think that's an issue. However, it may be good to just make sure.

You can safely do `B` without any drawbacks. I think if you do `B` it will be a clear win. `2 <= N + 1`, for all positive `N`.

I would benchmark method `C` against `B` to see if there are not performance drawbacks for production. If you want to be 100% safe. And if @matklad doesn't see using private api in those tests as drawback. I think `C` is a good option as well.

If you write `B` we will accept it right away.  

Personally I'm in favor of `C`, but it needs to be tested better. And I would like to hear feedback from @matklad. But I don't see any reasons why not to. Though I would like to see a benchmark that compares build time of production code between `B` and `C` if we decide to go with `C`.

**Comment by pmnoxx** [comment]
By moving tests from `near-network` -> `/tests` to `/src/tests`.
I see an a decrease of number of packages to build from `767` to `765`.

A quick check shows 89 files.
```
find . | grep "/tests" | grep "\.rs" | grep -v "\.git" | wc
```

That should decrease number of build targets from `767` to `678`. That is by 12%. I think that makes it a valid argument to go with `C`.

**Comment by pmnoxx** [comment]
@mzhangmzz Feel free to tag my in any future PRs or issue. I'm always happy to help.

**Comment by pmnoxx** [comment]
One note here. Sometimes we use "/src/tests" already. I would move tests into some other folder. For example "/src/integration_tests"

**Comment by pmnoxx** [comment]
@mzhangmzz Here is  proof of concept. Let's see if people like it. 
I do !

**Comment by pmnoxx** [comment]
@matklad  @mzhangmzz  It looks like moving fine from `/tests` to `/src/tests` doesn't require you us to change imports in any other file.

I'm conviced. It's a good change. Here is an example: https://github.com/near/nearcore/pull/5398/files

Let me know if you want to do it, or whenever I should do it. It should be a simple change. If you don't have time, I can do it on `Monday`. Fell free to take the change, in case you wan to ship it. It's up to you. It's your idea!

**Comment by mzhangmzz** [comment]
@pmnoxx I think whether a test should be put as unit tests or integration tests depends on what API they are testing (see @matklad comment here https://github.com/near/nearcore/pull/5343#issuecomment-974008963). In our codebase, I think that most tests access internal state instead of testing the public API, so they should be put as unit tests. 

Feel free to take it over. This is not my idea to be fair, it was all due to @matklad's comment on my PR and his awesome post.

**Comment by pmnoxx** [comment]
> @pmnoxx I think whether a test should be put as unit tests or integration tests depends on what API they are testing (see @matklad comment here [#5343 (comment)](https://github.com/near/nearcore/pull/5343#issuecomment-974008963)). In our codebase, I think that most tests access internal state instead of testing the public API, so they should be put as unit tests.

Ok, great. I think this is a good change. 

> Feel free to take it over. This is not my idea to be fair, it was all due to @matklad's comment on my PR and his awesome post.

I just wanted to make sure we don't duplicated the work. In case you already had PR's written. 

I included `PR` for doing it for `/chain`. Once (https://github.com/near/nearcore/pull/5398) gets approved I'll do it for other places.

**Comment by matklad** [comment]
>The document you mentioned proposes method B, you are proposing method C.

To clarify, the advice is "for crates.io libraries, go with B, for applications, go with C". 

>Also build time of C may theoretically make regular build slower, as there are more files to parse for regular build.

The cool thing that it doesn't. When the compiler sees `#[cfg(test)] mod tests;` and you are not building the tests specifically, the compiler completely skips `tests.rs` file (eg, it can even contain syntax errors). 

In other words, I feel strongly that we should go with C.

**Comment by pmnoxx** [comment]
> > The document you mentioned proposes method B, you are proposing method C.
> 
> To clarify, the advice is "for crates.io libraries, go with B, for applications, go with C".
> 
> > Also build time of C may theoretically make regular build slower, as there are more files to parse for regular build.
> 
> The cool thing that it doesn't. When the compiler sees `#[cfg(test)] mod tests;` and you are not building the tests specifically, the compiler completely skips `tests.rs` file (eg, it can even contain syntax errors).
> 
> In other words, I feel strongly that we should go with C.

Thanks, explanation. 

Let's first do it for `/chain`  then for the rest of the code base.

**Comment by mina86** [comment]
FYI, this broke expensive tests on NayDuck. NayDuck runs `cargo build --tests --target-dir target_expensive --features=expensive_tests` and expects test binaries to end up in `target_expensive/debug/deps` but now some of the executables are no longer there.

**Comment by pmnoxx** [comment]
@mina86 Thanks for fixing the issue. I see that we will have to make the same changes once we migrate `integration-tests` as well.

**Comment by matklad** [comment]
We also have some needless nested modules there: 

https://github.com/near/nearcore/blob/bd3529dd518e8080b19a43c4bb66a3b10f10d6c4/chain/chain/src/tests/doomslug.rs#L1-L3

Would be nice (for purely aesthetic reasons) to replace that with 

```rust
#![cfg(feature = "expensive_tests")] 
```

and unindent everything. Which again would require changing nayduck.
